### PR TITLE
David nav

### PR DESCRIPTION
**Pull request by DavidTTran**
Completed
- User stories 2-9, 13-16
- Add roles to users
- Update navigation based on roles
- Update register redirects based on roles
- Users can login and session is maintained

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    if current_user.role == 1
+      flash[:notice] = "You're already logged in."
+      redirect_to '/profile'
+    elsif current_user.role == 2
+      flash[:notice] = "You're already logged in."
+      redirect_to '/merchant/dashboard'
+    elsif current_user.role == 3
+      flash[:notice] = "You're already logged in."
+      redirect_to '/admin/dashboard'
+    end
```
Comment (Maxwell-Baird) [inline app/controllers/sessions_controller.rb:13]: If we have time to refactor do you think we can make a helper method based on this and the if block in SessionsController create

File: app/controllers/register_controller.rb [hunk 2]
```diff
@@ -5,13 +5,16 @@ def new
 
   def create
     @user = User.new(user_params)
+    @user.role = 1
```
Comment (DavidTTran) [inline app/controllers/register_controller.rb:8]: By default visitors have a role of 0, so when we're creating a new user - we have to assign the role to 1

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
```
Comment (DavidTTran) [inline app/controllers/sessions_controller.rb:14]: This looks ugly.. and long.. especially for a controller

File: app/controllers/cart_controller.rb [hunk 4]
```diff
@@ -29,4 +31,10 @@ def remove_item
   #   end
   #   redirect_to "/cart"
   # end
+
+  private
+
+  def check_role
+    render file: "/public/404" unless current_user.role < 3
```
Comment (colintalex) [inline app/controllers/cart_controller.rb:38]: Clean way of hiding the cart for admin, I like it

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      end
+    else
+      flash[:notice] = "Your credentials are incorrect."
+      render :new
+    end
+
+  end
+
+  def logout
+    session.clear
```
Comment (DavidTTran) [inline app/controllers/sessions_controller.rb:38]: This completely deletes the session.. But don't we want the session info to persist after logging out?

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -16,10 +16,31 @@
...
+        <%= link_to "Cart: #{cart.total_items}", "/cart" %>
+        <%= link_to "Profile", "/profile" %>
+        <%= link_to "Logout", "/logout" %>
+        <p>Logged in as: <%= current_user.name %></p>
+      <% elsif current_user.role == 3 %>
+        <%= link_to "Admin Dashboard", "/admin/dashboard" %>
+        <%= link_to "All Users", "/admin/users" %>
+        <%= link_to "Profile", "/profile" %>
+        <%= link_to "Logout", "/logout" %>
+        <p>Logged in as: <%= current_user.name %></p>
```
Comment (DavidTTran) [inline app/views/layouts/application.html.erb:42]: This.......

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+class SessionsController < ApplicationController
+  def new
+
+    if current_user.role == 1
+      flash[:notice] = "You're already logged in."
+      redirect_to '/profile'
+    elsif current_user.role == 2
```
Comment (colintalex) [inline app/controllers/sessions_controller.rb:7]: Clean and easy to read

File: db/migrate/20200407202208_change_password_to_password_digest.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class ChangePasswordToPasswordDigest < ActiveRecord::Migration[5.1]
+  def change
+    rename_column :users, :password, :password_digest
```
Comment (colintalex) [inline db/migrate/20200407202208_change_password_to_password_digest.rb:3]: For bcrypt

File: db/seeds.rb [hunk 9]
```diff
@@ -19,3 +20,9 @@
 #dog_shop items
 pull_toy = dog_shop.items.create(name: "Pull Toy", description: "Great pull toy!", price: 10, image: "http://lovencaretoys.com/image/cache/dog/tug-toy-dog-pull-9010_2-800x800.jpg", inventory: 32)
 dog_bone = dog_shop.items.create(name: "Dog Bone", description: "They'll love it!", price: 21, image: "https://img.chewy.com/is/image/catalog/54226_MAIN._AC_SL1500_V1534449573_.jpg", active?:false, inventory: 21)
+
+#users
+
+user = User.create(name: "User", address: "123 Test St", city: "Denver", state: "CO", zip: "80204", email: "user@example.com", password: "user", role: 1)
+merchant = User.create(name: "Merchant", address: "123 Test St", city: "Denver", state: "CO", zip: "80204", email: "merchant@example.com", password: "merchant", role: 2)
```
Comment (colintalex) [inline db/seeds.rb:27]: Thanks for adding these

**Review by colintalex — APPROVED** [review]
Looks good, lots of clean work on the session and nested paths

**Pull request merged**